pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+
+
+def mlfm_6(dmeas, c_1, c_2, c_3, c_4, c_5=0., c_6=0.):
+    r'''
+    Predict normalised LFM values from data in ``dmeas``.
+
+    The normalized LFM values are given by
+
+    .. math::
+
+        c_1 + c_2 (T_m - 25) + c_3 \log10(G_{POA}) + c_4 G_{POA}
+        + c_5 WS + c_6 / G_{POA}
+
+    where :math:`G_{POA}` is global plane-of-array (POA) irradiance in kW/m2,
+    :math:`T_m` is module temperature in C and :math:`WS` is wind speed in
+    m/s.
+
+    Parameters
+    ----------
+    dmeas : DataFrame
+        Must include columns:
+
+        * `'poa_global'` global plane of array irradiance. [W/m^2]
+        * `'temp_module'` module temperature. [C]
+
+        May include optional column:
+
+        * `'wind_speed'` wind speed [m/s].
+
+    c_1 : float
+        Constant term in model.
+    c_2 : float
+        Temperature coefficient in model. [1/C]
+    c_3 : float
+        Coefficient for low light log irradiance drop.
+    c_4 : float
+        Coefficient for high light linear irradiance drop.
+    c_5 : float, default 0
+        Coefficient for wind speed dependence.
+    c_6 : float, default 0
+        Coefficient for dependence on inverse irradiance.
+
+    Returns
+    -------
+    mlfm_6 : Series
+        Predicted values.
+
+    References
+    ----------
+    .. [1] Steve Ransome (SRCL) and Juergen Sutterlueti (Gantner Instruments)
+       "Quantifying Long Term PV Performance and Degradation under Real Outdoor
+       and IEC 61853 Test Conditions Using High Quality Module IV Measurements"
+       36th EU PVSEC, Marseille, France. September 2019
+     '''
+    mlfm_out = c_1 + c_2 * (dmeas['temp_module'] - T_STC) + \
+        c_3 * np.log10(dmeas['poa_global'] / G_STC) + \
+        c_4 * (dmeas['poa_global'] / G_STC) + \
+        c_6 / (dmeas['poa_global'] / G_STC)
+    if 'wind_speed' in dmeas.columns:
+        mlfm_out += c_5 * dmeas['wind_speed']
+    return mlfm_out
+
+
+def mlfm_fit(data, var_to_fit):
+    '''
+    Fit MLFM to data.
+
+    Parameters
+    ----------
+    data : DataFrame
+        Must include columns:
+
+        * 'poa_global' global plane of array irradiance. [W/m^2]
+        * 'temp_module' module temperature. [C]
+
+        Must include column named ``var_to_fit``.
+
+        May include optional column:
+
+            * 'wind_speed' wind speed [m/s].
+
+    var_to_fit : string
+        Column name in ``data`` containing variable being fit.
+
+    Returns
+    -------
+    pred : Series
+        Values predicted by the fitted model.
+
+    coeff : list
+        Model coefficients ``c_1`` to ``c_6``.
+
+    resid : Series
+        Residuals of the fitted model.
+
+    coeff_err : list
+        Standard deviation of error in each model coefficient.
+
+    See also
+    --------
+    mlfm_6
+    '''
+
+    # drop missing data
+    data = data.dropna()
+
+    c5_zero = 'wind_speed' not in data.columns
+    # if wind_speed is not present, add it
+    if c5_zero:
+        data['wind_speed'] = 0.
```

Comment:
Technically, c5 is _unidentifiable_ if the data do not contain windspeed information. As such, the algorithm may not converge robustly and may report convergence problems when you nonetheless keep c5 in the model. (Your bounds on c5 may temper the convergence issues to some degree.)

Unfortunately, `scipy.optimize.curve_fit` does not allow you to specify equal upper and lower bounds for c5 in this case, which here is presumably zero _by convention_. I can suggest a couple ways to work around this, but first I suggest examining the convergence info from `scipy.optimize.curve_fit` in a case where there is no windspeed info. (Do this by passing `full_output=True` and examining the  `mesg` and `ierint` entries.)

Reply:
I made comparison `curve_fit`s using datasets n05667_Y13_R1k6_fClear_041.csv and x19074001_iec61853_041.csv, both of which have no `wind_speed` info. For the comparison, I removed the `c5` coefficient altogether from `mlfm_6` (instead using a new `mlfm_5` function for `func`). The remaining coefficients appear to be stable between the two approaches without convergence errors in either (although the convergence criterion may change from `gtol` to `ftol`). I did not check the results of of the fit beyond this. It appears that it is ok to use `mlfm_6` in all cases, but I would still recommend checking for convergence.